AutoAttack does not run

```console
$ python -m mart experiment=CIFAR10_RobustBench trainer=gpu fit=false +trainer.limit_test_batches=1 +attack@model.modules.input_adv_test=classification_autoattack

[2023-02-16 11:35:14,513][mart.utils.utils][ERROR] -
Traceback (most recent call last):
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 92, in _call_target
    return _target_(*args, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'adversary'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/weilinxu/coder/MART/mart/utils/utils.py", line 54, in wrap
    metric_dict, object_dict = task_func(cfg=cfg)
  File "/home/weilinxu/coder/MART/mart/tasks/lightning.py", line 39, in lightning
    model: LightningModule = hydra.utils.instantiate(cfg.model)
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 222, in instantiate
    return instantiate_node(
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 334, in instantiate_node
    value = instantiate_node(
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 349, in instantiate_node
    dict_items[key] = instantiate_node(
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 339, in instantiate_node
    return _call_target(_target_, partial, args, kwargs, full_key)
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 97, in _call_target
    raise InstantiationException(msg) from e
hydra.errors.InstantiationException: Error in call to target 'mart.attack.adversary_wrapper.NormalizedAdversaryAdapter':
TypeError("__init__() got an unexpected keyword argument 'adversary'")
full_key: model.modules.input_adv_test
[2023-02-16 11:35:14,514][mart.utils.utils][INFO] - Closing loggers...
Error executing job with overrides: ['experiment=CIFAR10_RobustBench', 'trainer=gpu', 'fit=false', '+trainer.limit_test_batches=1', '+attack@model.modules.input_adv_test=classification_autoattack']
Error in call to target 'mart.attack.adversary_wrapper.NormalizedAdversaryAdapter':
TypeError("__init__() got an unexpected keyword argument 'adversary'")
full_key: model.modules.input_adv_test

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

--- a/mart/attack/adversary_wrapper.py
+++ b/mart/attack/adversary_wrapper.py
@@ -20,17 +20,15 @@ class NormalizedAdversaryAdapter(torch.nn.Module):
     External adversaries commonly take input of NCWH-[0,1] and return input_adv in the same format.
     """
 
-    def __init__(
-        self, external_adversary: Callable[[torch.Tensor, torch.Tensor, torch.nn.Module], None]
-    ):
+    def __init__(self, adversary: Callable[[torch.Tensor, torch.Tensor, torch.nn.Module], None]):
         """
 
         Args:
-            external_adversary (functools.partial): A partial of an adversary object which awaits model.
+            adversary (functools.partial): A partial of an adversary object which awaits model.
         """
         super().__init__()
 
-        self.adversary = external_adversary
+        self.adversary = adversary
 
     def forward(
         self,

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -13,7 +13,7 @@ adversary:
     version: custom
     attacks_to_run:
       - apgd-dlr
-  adapter:
+  wrapper:
     _target_: mart.utils.adapters.CallableAdapter
     _partial_: true
     redirecting_fn: run_standard_evaluation

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -1,7 +1,7 @@
 _target_: mart.attack.NormalizedAdversaryAdapter
 adversary:
-  _target_: mart.utils.adapters.CallableAdapter
-  instance:
+  _target_: mart.utils.adapters.PartialAdapter
+  partial:
     _target_: autoattack.AutoAttack
     _partial_: true
     norm: Linf
@@ -10,4 +10,6 @@ adversary:
     version: custom
     attacks_to_run:
       - apgd-dlr
-  redirecting_fn: run_standard_evaluation
+  adapter:
+    _target_: mart.utils.adapters.CallableAdapter
+    redirecting_fn: run_standard_evaluation

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -1,5 +1,5 @@
 _target_: mart.attack.NormalizedAdversaryAdapter
-external_adversary:
+adversary:
   _target_: mart.utils.adapters.CallableAdapter
   instance:
     _target_: autoattack.AutoAttack

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -1,6 +1,6 @@
 _target_: mart.attack.NormalizedAdversaryAdapter
 adversary:
-  _target_: mart.utils.adapters.PartialAdapter
+  _target_: mart.utils.adapters.PartialInstanceWrapper
   partial:
     _target_: autoattack.AutoAttack
     _partial_: true
@@ -10,6 +10,4 @@ adversary:
     version: custom
     attacks_to_run:
       - apgd-dlr
-  adapter:
-    _target_: mart.utils.adapters.CallableAdapter
-    redirecting_fn: run_standard_evaluation
+  redirecting_fn: run_standard_evaluation

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -10,4 +10,7 @@ adversary:
     version: custom
     attacks_to_run:
       - apgd-dlr
-  redirecting_fn: run_standard_evaluation
+  adapter:
+    _target_: mart.utils.adapters.CallableAdapter
+    _partial_: true
+    redirecting_fn: run_standard_evaluation

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -1,5 +1,5 @@
 _target_: mart.attack.NormalizedAdversaryAdapter
-adversary:
+external_adversary:
   _target_: mart.utils.adapters.CallableAdapter
   instance:
     _target_: autoattack.AutoAttack

--- a/mart/configs/attack/classification_autoattack.yaml
+++ b/mart/configs/attack/classification_autoattack.yaml
@@ -4,6 +4,9 @@ adversary:
   partial:
     _target_: autoattack.AutoAttack
     _partial_: true
+    # AutoAttack needs to specify device for PyTorch tensors: cpu/cuda
+    # We can not use ${trainer.accelerator} because the vocabulary is different: cpu/gpu
+    # device: cpu
     norm: Linf
     # 8/255
     eps: 0.03137254901960784

--- a/mart/utils/adapters.py
+++ b/mart/utils/adapters.py
@@ -43,18 +43,18 @@ class CallableAdapter:
 class PartialInstanceWrapper:
     """Make a partial class object callable."""
 
-    def __init__(self, partial: Callable, adapter):
+    def __init__(self, partial: Callable, wrapper: Callable):
         """
 
         Args:
             partial (Callable): A partial of a class object.
-            adapter (CallableAdapter): An adapter that creates the `__call__` method.
+            adapter (Callable): An adapter that creates the `__call__` method.
         """
         self.partial = partial
-        self.adapter = adapter
+        self.wrapper = wrapper
 
     def __call__(self, *args: Any, **kwargs: Any) -> Callable:
         # Turn a partial to a class object.
         instance = self.partial(*args, **kwargs)
         # Make the object callable.
-        return self.adapter(instance)
+        return self.wrapper(instance)

--- a/mart/utils/adapters.py
+++ b/mart/utils/adapters.py
@@ -5,41 +5,51 @@
 # agreement between Intel Corporation and you.
 #
 
-import functools
+from typing import Any, Callable
 
-__all__ = ["CallableAdapter"]
+__all__ = ["CallableAdapter", "PartialAdapter"]
 
 
 class CallableAdapter:
     """Adapter to make an object callable."""
 
-    def __init__(self, instance, redirecting_fn):
+    def __init__(self, redirecting_fn):
+        """
+
+        Args:
+            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
+        """
+        assert redirecting_fn != ""
+
+        self.redirecting_fn = redirecting_fn
+
+    def __call__(self, instance):
         """
 
         Args:
             instance (object): instance to make callable.
-            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
         """
-        assert instance is not None
-        assert redirecting_fn != ""
 
-        self.instance = instance
-        self.redirecting_fn = redirecting_fn
+        function = getattr(instance, self.redirecting_fn)
+        assert callable(function)
+        return function
 
-    def __call__(self, *args, **kwargs):
+
+class PartialAdapter:
+    """Make a partial class object callable."""
+
+    def __init__(self, partial: Callable, adapter: Callable):
         """
 
         Args:
-            args (Any): values to use in the callable method.
-            kwargs (Any): keyword values to use in the callable method.
+            partial (Callable): A partial of a class object.
+            adapter (Callable): An adapter to make an object callable.
         """
-        # First make a partial to a function.
-        if isinstance(self.instance, functools.partial):
-            self.instance = self.instance(*args, **kwargs)
-            return self
+        self.partial = partial
+        self.adapter = adapter
 
-        function = getattr(self.instance, self.redirecting_fn)
-
-        assert callable(function)
-
-        return function(*args, **kwargs)
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        # Turn a partial to a class object.
+        instance = self.partial(*args, **kwargs)
+        # Make the object callable.
+        return self.adapter(instance)

--- a/mart/utils/adapters.py
+++ b/mart/utils/adapters.py
@@ -7,49 +7,54 @@
 
 from typing import Any, Callable
 
-__all__ = ["CallableAdapter", "PartialAdapter"]
+__all__ = ["CallableAdapter", "PartialInstanceWrapper"]
 
 
 class CallableAdapter:
     """Adapter to make an object callable."""
 
-    def __init__(self, redirecting_fn):
-        """
-
-        Args:
-            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
-        """
-        assert redirecting_fn != ""
-
-        self.redirecting_fn = redirecting_fn
-
-    def __call__(self, instance):
+    def __init__(self, instance, redirecting_fn):
         """
 
         Args:
             instance (object): instance to make callable.
+            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
+        """
+        assert instance is not None
+        assert redirecting_fn != ""
+
+        self.instance = instance
+        self.redirecting_fn = redirecting_fn
+
+    def __call__(self, *args, **kwargs):
         """
 
-        function = getattr(instance, self.redirecting_fn)
+        Args:
+            args (Any): values to use in the callable method.
+            kwargs (Any): keyword values to use in the callable method.
+        """
+        function = getattr(self.instance, self.redirecting_fn)
+
         assert callable(function)
-        return function
+
+        return function(*args, **kwargs)
 
 
-class PartialAdapter:
+class PartialInstanceWrapper:
     """Make a partial class object callable."""
 
-    def __init__(self, partial: Callable, adapter: Callable):
+    def __init__(self, partial: Callable, redirecting_fn: str):
         """
 
         Args:
             partial (Callable): A partial of a class object.
-            adapter (Callable): An adapter to make an object callable.
+            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
         """
         self.partial = partial
-        self.adapter = adapter
+        self.redirecting_fn = redirecting_fn
 
-    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+    def __call__(self, *args: Any, **kwargs: Any) -> Callable:
         # Turn a partial to a class object.
         instance = self.partial(*args, **kwargs)
         # Make the object callable.
-        return self.adapter(instance)
+        return CallableAdapter(instance, self.redirecting_fn)

--- a/mart/utils/adapters.py
+++ b/mart/utils/adapters.py
@@ -43,18 +43,18 @@ class CallableAdapter:
 class PartialInstanceWrapper:
     """Make a partial class object callable."""
 
-    def __init__(self, partial: Callable, redirecting_fn: str):
+    def __init__(self, partial: Callable, adapter):
         """
 
         Args:
             partial (Callable): A partial of a class object.
-            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
+            adapter (CallableAdapter): An adapter that creates the `__call__` method.
         """
         self.partial = partial
-        self.redirecting_fn = redirecting_fn
+        self.adapter = adapter
 
     def __call__(self, *args: Any, **kwargs: Any) -> Callable:
         # Turn a partial to a class object.
         instance = self.partial(*args, **kwargs)
         # Make the object callable.
-        return CallableAdapter(instance, self.redirecting_fn)
+        return self.adapter(instance)

--- a/mart/utils/adapters.py
+++ b/mart/utils/adapters.py
@@ -5,6 +5,8 @@
 # agreement between Intel Corporation and you.
 #
 
+import functools
+
 __all__ = ["CallableAdapter"]
 
 
@@ -31,6 +33,11 @@ class CallableAdapter:
             args (Any): values to use in the callable method.
             kwargs (Any): keyword values to use in the callable method.
         """
+        # First make a partial to a function.
+        if isinstance(self.instance, functools.partial):
+            self.instance = self.instance(*args, **kwargs)
+            return self
+
         function = getattr(self.instance, self.redirecting_fn)
 
         assert callable(function)

--- a/tests/test_experiments.py
+++ b/tests/test_experiments.py
@@ -146,6 +146,25 @@ def test_cifar10_cnn_experiment(classification_cfg, tmp_path):
 
 
 @RunIf(sh=True)
+def test_cifar10_cnn_autoattack_experiment(classification_cfg, tmp_path):
+    """Test CIFAR10 CNN AutoAttack experiment."""
+    overrides = classification_cfg["datamodel"]
+    command = [
+        module,
+        "-m",
+        "experiment=CIFAR10_CNN",
+        "hydra.sweep.dir=" + str(tmp_path),
+        "++datamodule.train_dataset.image_size=[3,32,32]",
+        "++datamodule.train_dataset.num_classes=10",
+        "fit=false",
+        "+attack@model.modules.input_adv_test=classification_autoattack",
+        '+model.modules.input_adv_test.adversary.partial.device="cpu"',
+        "+trainer.limit_test_batches=1",
+    ] + overrides
+    run_sh_command(command)
+
+
+@RunIf(sh=True)
 def test_cifar10_robust_bench_experiment(classification_cfg, tmp_path):
     """Test CIFAR10 Robust Bench experiment."""
     overrides = classification_cfg["trainer"] + classification_cfg["datamodel"]

